### DISCUSSION TITLE
Questions resulting from (mis)using solid-js in node "backend" code

### DISCUSSION THREAD

**Discussion by danieltroger** [solidjs/solid#1465] [Q&A]
So I'm running solid-js in node, not to render HTML but just to have the reactive system (signals, memos and effects) since it is very powerful. (To do that I've had to map `solid-js` imports to `solid-js/dist/solid.js`)

My application is a raspberry pi that has a couple sensors and inputs and then controls outputs according to it.

I've come to wonder, now that there's no "rendering phase" (I assume, since no HTML is rendered?), if there's any difference between `createComputed`, `createRenderEffect` or `createEffect` for me? Does it make sense to have functions that change state in `createComputed` and then the code that actually updates the GPIO output in a `createEffect` to kind of "batch" all computations and only update the outputs with the final result?

**Comment by samuelstroschein** [comment]
I can't give an answer to your question, but I am curious how you leverage Solid's reactive system in an application running on a Raspberry PI. 

Do you care to elaborate?

**Reply by danieltroger** [reply]
Hi @samuelstroschein,

The application is a program running in nodejs that controls the heating of a house.
It has a couple inputs:
1. Three thermometers
2. Two capacitive sensors in the pellets stove
3. A motor protection for a feeder motor (which feeds from another building containing pellets into the smaller "storage" inside of the pellets stove)
4. A camera pointed at the display of the stove

And outputs:
1. Whether the stove should be on or not
2. How many kw the store should burn at
3. Whether the feeder should feed pellets into the stove
5. Thermometer power (they are digital and get stuck sometimes so I can programmatically reboot them that way)
6. Whether a fallback electric water heater should run

Then it also has a websocket server that a frontend connects to where I plan to do things like "bath mode" where the water is kept at a hotter temperature than usual and some configuration and monitoring.

Using the I/O it does a couple things like feeding pellets when it should if the motor protection hasn't triggered, turning the stove on when the water is too cold, regulating the power depending on how fast the temperature rises (I yet have to implement this part), turning the stove off when it's warm enough. Then it takes pictures using the camera for debugging purposes, stores the temperatures and sends them to the frontend if someone is there, etc.

Anyways, inside of this program it's very useful to have reactivity IMO. One reason being that it's just convenient: I can set up the I/O in one part of the program and then easily have multiple parts of it react to I/O changes - for example sending temperature both to the frontend, using it to calculate power of the stove and saving it to a file just requires three effects that depend on the temperature signal.

But also since a previous version of the program always had to be restarted when changing some config and then lost some state, which annoyed me (i.e. it "forgot" that the stove was on and then re-started it just because I wanted to change the target temperature) and by making the config values signals it becomes insanely easy to just change them from the frontend and then everything that depends on the config values will instantly update.

I now can even just add and remove thermometers dynamically  (and they will start/stop being read) since I'm using solid's `For` component to start a "thermometer reader" (which actually works on the server and can even return whatever one wants if one ts-ignores the typing).

Hope this helps!

**Reply by samuelstroschein** [reply]
This helps and is interesting. Did you consider alternatives to Solid's reactive system in your Raspberry PI application?

**Reply by danieltroger** [reply]
Only my own which I had created before knowing about solid-js. The idea for that was basically an "Infinite promise" so like promises but that you can resolve them multiple times. Then listen on it using `for await`. The issue is that it always takes a tick in the microtask queue from updating a value to the updates propagating and that's annoying to deal with in some cases, so I went with solid-js which seems better thought out.

Are there any other alternatives which you think I should have considered?

**Reply by samuelstroschein** [reply]
> Are there any other alternatives which you think I should have considered? 

Not that I am aware of. My interest stems from my curiosity about where SolidJS's reactive system could be/is used that has nothing to do with UI. You have such a use case. Super interesting, thanks for sharing!
